src/common/base/NebulaKeyUtils.cpp

```diff
@@ -88,6 +88,59 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
     return key;
 }
 
+// static
+void NebulaKeyUtils::indexRaw(const IndexValues &values, std::string& raw) {
+    std::vector<int32_t> colsLen;
+    for (auto& col : values) {
+        if (col.first == cpp2::SupportedType::STRING) {
+            colsLen.emplace_back(col.second.size());
+        }
+        raw.append(col.second.data(), col.second.size());
+    }
+    for (auto len : colsLen) {
+        raw.append(reinterpret_cast<const char*>(&len), sizeof(int32_t));
+    }
+}
+
+// static
+std::string NebulaKeyUtils::vertexIndexKey(PartitionID partId, IndexID indexId, VertexID vId,
+                                           const IndexValues& values) {
+    int32_t item = (partId << 8) | static_cast<uint32_t>(NebulaKeyType::kIndex);
+    std::string key;
+    key.reserve(256);
+    key.append(reinterpret_cast<const char*>(&item), sizeof(int32_t))
+       .append(reinterpret_cast<const char*>(&indexId), sizeof(IndexID));
+    indexRaw(values, key);
+    key.append(reinterpret_cast<const char*>(&vId), sizeof(VertexID));
+    return key;
+}
+
+// static
+std::string NebulaKeyUtils::edgeIndexKey(PartitionID partId, IndexID indexId,
+                                         VertexID srcId, EdgeRanking rank,
+                                         VertexID dstId, const IndexValues& values) {
+    int32_t item = (partId << 8) | static_cast<uint32_t>(NebulaKeyType::kIndex);
+    std::string key;
+    key.reserve(256);
+    key.append(reinterpret_cast<const char*>(&item), sizeof(int32_t))
+       .append(reinterpret_cast<const char*>(&indexId), sizeof(IndexID));
+    indexRaw(values, key);
+    key.append(reinterpret_cast<const char*>(&srcId), sizeof(VertexID))
```

Comment:
I am sorry for my poor understanding
1. Why we don't need to encode edge type here?
2. Why we encode the `edge key` in index key instead of index value?
3. I guess we can scan all index of first property, am I right?

Reply:
> I am sorry for my poor understanding
> 
> 1. Why we don't need to encode edge type here?
IndexId can  identify the `edgetype` or `tagId`
> 2. Why we encode the `edge key` in index key instead of index value?
we only need scan key by prefix or range, so we need encode the values into key , I don't know if my answer is clear. 
> 3. I guess we can only scan all index of first property in IndexValues, am I right?
Yes，you are right, if is a range query ,only scan first property. if is a equivalent query, can be first N properties.

Reply:
Maybe the second one need a little more hint.

Reply:
> Maybe the second one need a little more hint.

for second one, the index key format is :
`partId + indexId + propValues + srcId + rank + dstId . `
the propValues format is : 
`edgeType.col1 + edgeType.col2 + degeType.col3+... `. the column value is encoded(int and timestamp is big-endain).